File: src/common/actor/Level.java
package common.actor;

import javafx.scene.canvas.Canvas;

import java.util.*;

/**
 * A simple level structure.
 *
 * @author Alex
 */
abstract public class Level extends Observable implements Observer, ILevel {

    private BackgroundImage        backgroundImage = new BackgroundImage();
    private List<Actor>            npcs            = Collections.synchronizedList( new ArrayList<>() );
    private List<ControlableActor> players         = Collections.synchronizedList( new LinkedList<>() );
    private List<LevelElement>     levelElements   = Collections.synchronizedList( new ArrayList<>() );
    private List<Collectable>      collectables    = Collections.synchronizedList( new ArrayList<>() );

    public Level( Canvas gameCanvas ) {
        reset( gameCanvas );
    }

    private void addCollision() {
        synchronized ( this.levelElements ) {
            this.levelElements.forEach(
                    levelElement -> players.forEach(
                            pacMan -> pacMan.addCollidingActor( levelElement ) ) );
        }
    }

    private void addCollectables() {
        synchronized ( this.collectables ) {
            this.collectables.forEach(
                    collectable -> players.forEach(
                            player -> player.addCollidingActor( collectable ) ) );
        }
    }

    protected boolean addLevelElement( Canvas gameCanvas, final LevelElement levelElement ) {
        boolean[] xy = CollisionCheck.isInBounds( levelElement, gameCanvas );
        if ( xy[ 0 ] && xy[ 1 ] ) {
            return addLevelElement( levelElement );
        }
        return false;
    }

    @Override
    public synchronized void deleteObservers() {
        this.backgroundImage.deleteObservers();
        for ( Actor npc : this.npcs ) {
            npc.deleteObservers();
        }
        for ( ControlableActor player : this.players ) {
            player.deleteObservers();
        }
        for ( LevelElement levelElement : this.levelElements ) {
            levelElement.deleteObservers();
        }
        for ( Collectable collectable : this.collectables ) {
            collectable.deleteObservers();
        }
        super.deleteObservers();
    }

    @Override
    public synchronized void render( Canvas canvas, int fps ) {
        try {
            this.backgroundImage.draw( canvas );
            this.draw( npcs, canvas );
            this.draw( levelElements, canvas );
            this.draw( collectables, canvas );
            this.draw( players, canvas );
        }
        catch ( ConcurrentModificationException cme ) {
            cme.printStackTrace();
        }
    }

    private synchronized <T extends Actor> void draw( final List<T> list, Canvas canvas ) {
        int size = list.size();
        for ( int i = 0 ; i < size ; i++ ) {
            final T t = list.get( i );
            if ( t != null ) {
                t.draw( canvas );
            }
            size = list.size();
        }
    }

    protected boolean collidesWithLevelElement( final Actor a ) {
        for ( LevelElement levelElement : getLevelElements() ) {
            if ( levelElement.doesCollide( a ) ) {
                return true;
            }
        }
        return false;
    }


    protected boolean collidesWithPlayer( final Actor d ) {
        for ( ControlableActor controlableActor : getPlayers() ) {
            if ( controlableActor.doesCollide( d ) ) {
                return true;
            }
        }
        return false;
    }

    protected boolean collidesWithCollectable( final Actor a ) {
        for ( Collectable coll : getCollectables() ) {
            if ( coll != null ) {
                if ( coll.doesCollide( a ) ) {
                    return true;
                }
            }
        }
        return false;
    }

    protected boolean collected( final Collectable collectable ) {
        collectable.deleteObservers();
        final List<ControlableActor> players = Collections.synchronizedList( this.players );
        synchronized ( players ) {
            int size = players.size();
            for ( int i = 0 ; i < size ; i++ ) {
                final ControlableActor c = players.get( i );
                c.removeCollisionActor( collectable );
                size = players.size();
            }
        }
        final boolean result = getCollectables().remove( collectable );
        this.getCollectables().retainAll( getCollectables() );
        return result;
    }

    protected boolean addCollectable( final Collectable c ) {
        c.addObserver( this );
        final List<ControlableActor> players = Collections.synchronizedList( this.players );
        synchronized ( players ) {
            int size = players.size();
            for ( int i = 0 ; i < size ; i++ ) {
                players.get( i ).addCollidingActor( c );
                size = players.size();
            }
            this.players = players;
        }
        return this.collectables.add( c );
    }

    protected boolean addPlayer( ControlableActor player ) {
        return this.players.add( player );
    }

    protected boolean addLevelElement( LevelElement levelElement ) {
        return this.levelElements.add( levelElement );
    }

    private void initializeMembers() {
        backgroundImage = new BackgroundImage();
        npcs = new ArrayList<>();
        players = new LinkedList<>();
        levelElements = new ArrayList<>();
        collectables = new ArrayList<>();
    }

    public void reset( Canvas gameCanvas ) {
        initializeMembers();
        createLevel( gameCanvas );
        combineResources();
    }

    protected void combineResources() {
        addCollision();
        addCollectables();
    }

    protected abstract boolean isGameFinished();

    protected List<ControlableActor> getPlayers() {
        return players;
    }

    protected List<LevelElement> getLevelElements() {
        return levelElements;
    }

    public List<Actor> getNpcs() {
        return npcs;
    }

    protected List<Collectable> getCollectables() {
        return this.collectables;
    }

    protected void setBackgroundImage( final String filepath, final int width, final int height ) {
        this.backgroundImage.setCurrentImage( filepath );
        this.backgroundImage.setWidth( width );
        this.backgroundImage.setHeight( height );
    }
}

File: Common/Actor/src/common/actor/Level.kt
package common.actor

import javafx.scene.canvas.Canvas
import java.util.*

/**
 * A simple level structure.
 *
 * @author Alex
 */
abstract class Level(gameCanvas: Canvas) : Observable(), Observer, ILevel {

    private var backgroundImage = BackgroundImage()
    private var npcs = mutableListOf<Actor>()
    var players = mutableListOf<ControllableActor>()
        private set

    private var levelElements = mutableListOf<LevelElement>()
    protected var collectables = mutableListOf<Collectable>()
        private set


    protected fun addLevelElement(gameCanvas: Canvas,
                                  levelElement: LevelElement): Boolean =
            if (CollisionCheck.isInBounds(levelElement, gameCanvas)) {
                addLevelElement(levelElement)
            } else false

    @Synchronized
    override fun deleteObservers() {
        backgroundImage.deleteObservers()
        npcs.forEach { npc -> npc.deleteObservers() }
        players.forEach { player -> player.deleteObservers() }
        levelElements.forEach { levelElement -> levelElement.deleteObservers() }
        collectables.forEach { collectable -> collectable.deleteObservers() }
        super.deleteObservers()
    }

    @Synchronized
    override fun render(canvas: Canvas, fps: Int) {
        backgroundImage.draw(canvas)
        draw(npcs, canvas)
        draw(levelElements, canvas)
        draw(collectables, canvas)
        draw(players, canvas)
    }

    private fun <T : Actor> draw(list: List<T>, canvas: Canvas) {
        val size = list.size
        for (i in 0 until size) {
            list[i].draw(canvas)
        }
    }

    protected infix fun collidesWithLevelElement(a: Actor): Boolean = levelElements.any { it.doesCollide(a) }

    protected infix fun collidesWithPlayer(d: Actor): Boolean = players.any { it.doesCollide(d) }

    protected infix fun collidesWithCollectable(a: Actor): Boolean = collectables.any { it.doesCollide(a) }

    protected infix fun collected(collectable: Collectable): Boolean {
        collectable.deleteObservers()
        val list = Collections.synchronizedList(
                players)
        synchronized(list) {
            for (player in list) {
                player.removeCollisionActor(collectable)
            }
        }
        val result = collectables.remove(collectable)
        collectables.retainAll(collectables)
        return result
    }

    protected open fun addCollectable(c: Collectable): Boolean {
        c.addObserver(this)
        val list = Collections.synchronizedList(players)
        for (player in list) {
            player.addCollidingActor(c)
        }
        players = list
        return collectables.add(c)
    }

    protected infix fun addPlayer(player: ControllableActor): Boolean = players.add(player)

    protected open infix fun addLevelElement(levelElement: LevelElement): Boolean = levelElements.add(levelElement)

    private fun reset(gameCanvas: Canvas) {
        fun addCollision() = levelElements.forEach { levelElement: LevelElement ->
                    players.forEach { pacMan: ControllableActor ->
                        pacMan addCollidingActor levelElement
                    }
                }

        fun addCollectables() = collectables.forEach { collectable: Collectable ->
                    players.forEach { player: ControllableActor ->
                        player addCollidingActor collectable
                    }
                }

        backgroundImage = BackgroundImage()
        npcs = mutableListOf()
        players = mutableListOf()
        levelElements = mutableListOf()
        collectables = mutableListOf()
        createLevel(gameCanvas)
        addCollision()
        addCollectables()
    }

    open fun resetAndThan(gameCanvas: Canvas, thanDo: (Canvas) -> Unit) {
        this.reset(gameCanvas)
        thanDo(gameCanvas)
    }

    protected abstract val isGameFinished: Boolean

    protected fun setBackgroundImage(filepath: String,
                                     width: Int,
                                     height: Int) {
        backgroundImage.currentImage = TextureBuffer.loadImage(filepath)
        backgroundImage.width = width.toDouble()
        backgroundImage.height = height.toDouble()
    }

    init {
        reset(gameCanvas)
    }
}
